# sourced by 'tcga_mirna_ui.R'
# save as 'tcga_mirna_help.R'
# help file for tcga_mirna ui part

fluidRow(
  style = "width:85%",
  column(
    width = 12,
    # collapsable panel
    shiny::tags$div(
      class = "panel panel-default",
      shiny::tags$div(
        class = "panel-heading",
        shiny::tags$h3(
          class = "panel-title text-left",
          HTML('<a data-toggle="collapse" href="#sqTCGA_mirna">
<i class="fa fa-question fa-fw"></i> 
               Click here for help</a>')
        )
      ),
      shiny::tags$div(
        id = "sqTCGA_mirna", class = "panel-collapse collapse",
        shiny::tags$div(
          class = "panel-body",
          # here comes the content
          ###############
          column(
            width = 12,
            # overall description----
            column(
              width = 10, offset = 1,

              shiny::tags$h3("What is miRNA?", class = "text-success"),

              shiny::tags$p(
                class = "text-justify", "A microRNA (abbreviated miRNA) is a small non-coding RNA molecule (containing about 22 nucleotides) found in plants, animals and some viruses, that functions in RNA silencing and post-transcriptional regulation of gene expression. See details at: ",
                shiny::tags$a(href = "http://www.sciencedirect.com/science/article/pii/S0092867404000455?via%3Dihub", 
                              shiny::tags$span("D. P. Bartel, MicroRNAs: Genomics, Biogenesis, Mechanism, and Function. Cell. 116, 281â€“297 (2004)",style="color:#008176"),target = "_blank"), "."
              ),
              
              shiny::tags$hr(width = "100%")
            ),
            column(
              width = 10, offset = 1,

              shiny::tags$h3("miRNA-gene data source", class = "text-success"),

              shiny::tags$p(
                class = "text-justify", "miRNA regulation data was collected form databases: include experimental verified (papers, ",
                shiny::tags$a(href = "http://diana.imis.athena-innovation.gr/DianaTools/index.php?r=site/index", 
                              shiny::tags$span("TarBase, ",style="color:#008176"),target = "_blank"), 
                shiny::tags$a(href = "http://mirtarbase.mbc.nctu.edu.tw/php/index.php", 
                              shiny::tags$span("miRTarBase, ",style="color:#008176"),target = "_blank"),
                shiny::tags$a(href = "http://www.mir2disease.org/", 
                              shiny::tags$span("mir2disease",style="color:#008176"),target = "_blank"),"), and ",
                shiny::tags$a(href = "http://www.targetscan.org/vert_70/", 
                              shiny::tags$span("targetscan, ",style="color:#008176"),target = "_blank"),
                shiny::tags$a(href = "http://34.236.212.39/microrna/home.do", 
                              shiny::tags$span("miRanda",style="color:#008176"),target = "_blank"), 
                " predicted. And only miRNA-gene pairs who have been recorded will be used to calculate a expression correlation here."
              ),
              # shiny::tags$p(class="text-justify",),

              shiny::tags$hr(width = "100%")
            ),
            column(
              width = 10, offset = 1,

              shiny::tags$h3("Method", class = "text-success"),

              shiny::tags$p(class = "text-justify", "1. Correlation was calculated between gene and miRNA expression in all samples , in consideration of the presence of positive regulators like transcription factors, a miRNA-gene pair with correlation coefficient <0.5 and p.value<0.05 will be considered as a potential negatively regulation pair."),
              shiny::tags$p(class = "text-justify", "2. Only the miRNA-gene pairs who have been recorded in databases we refered below will be calculated at step 1."),
              shiny::tags$p(class = "text-justify", "3. Networks are generated by R packages: networkD3 and visNetwork. Subgroups of networkD3 is generated by getting communities in a graph via random walks in igraph R package."),
              # shiny::tags$p(class="text-justify",),

              shiny::tags$hr(width = "100%")
            ),
            # result description----
            column(
              width = 10, offset = 1,
              shiny::tags$h3("Results commentary", class = "text-success"),
              shiny::tags$table(
                class = "table table-striped",

                shiny::tags$thead(
                  shiny::tags$th("Results type",width=4),
                  shiny::tags$th("Description",width=8)
                ),
                shiny::tags$tbody(
                  shiny::tags$tr(
                    shiny::tags$td("networkD3"),
                    shiny::tags$td("The networkD3 R package is used to generate this beautiful network, if you need to get a cool miRNA regulation network for your paper, it will be a nice choice. In this network, node size is positively correlate to the node's degree, which is simplely defined by the number of nodes it connect to, for example, when miRNA regulate n genes, it's degree is n, and node size is sqrt(n)+6. Subgroups  are generated by getting communities in a graph via random walks by igraph R package, and shown by different colors. ")
                    ),
                  shiny::tags$tr(
                  shiny::tags$td("visNetwork"),
                  shiny::tags$td("The visNetwork R package is used to generate this useful network, a node represent a miRNA or gene, an edge represent a regulation of miRNA to gene. We have cluster them by color, and you can click, drag, delele and add nodes or edges in the network as you wish. Node size is positively correlate to the node's degree similar to networkD3, and edge width is defined by absolute value of (correlation coefficient-1).")
                )
                )
              )
            ) # result description end
          ) # all description end
        )
      )
    )
  )
)
